### DISCUSSION TITLE
How to synchronously add or remove LineCartesianLayer.Line

### DISCUSSION THREAD

**Discussion by Tyler-Lopez** [patrykandpatrick/vico#924] [Questions] [answered]
### Question

Hi there!

I recently opened a [bug issue](https://github.com/patrykandpatrick/vico/issues/914) that was closed. It involved a crash when dynamically adding and removing `LineCartesianLayer`. The issue was closed as the root of the problem was a failure of Compose API—not Vico. However, @Gowsky mentioned within that issue that there is already a functional way to add or remove `LineCartesianLayer.Line` (not `LineCartesianLayer`) synchronously using the `ExtraStore`.

It was mentioned that @PatrykGoworowski would color in some detail as to how this would be done. It has been a little while, and I thought - rather than bump that useful solution within the issue that I would make a separate question so that others could easily find this.

Any thoughts here? 

This is an example of doing the dynamic addition and removal incorrectly, without using `ExtraStore`. https://gist.github.com/Tyler-Lopez/a13b5efa6235182b6862c8f69cae63bc

Thanks!

### UI framework(s)

Jetpack Compose

**Comment by patrickmichalik** [comment, marked as answer]
Hello! As described [in the wiki](https://patrykandpatrick.com/vico/wiki/cartesian-charts/data#asynchrony-and-extras), the APIs that facilitate synchronized setup provide access to the `CartesianChartModel` or derived data (usually the `ExtraStore`) via function parameters. It can be observed that `LineCartesianLayer.LineProvider.series` doesn’t fit this criterion. (Still, there are valid use cases for reinvoking `LineCartesianLayer.Line.series` with new `LineCartesianLayer.Line`s—consider a user-accessible color setting.)

While `LineCartesianLayer.LineProvider.series` doesn’t facilitate synchronized setup, `LineCartesianLayer.LineProvider` itself does—the `getLine` function has an `ExtraStore` parameter. By creating a custom implementation of the interface and using extras, you can make `LineCartesianLayer.Line`-related changes in a synchronized fashion.

```kt
abstract fun getLine(seriesIndex: Int, extraStore: ExtraStore): LineCartesianLayer.Line
```

For completeness, it should be noted that in the particular example provided, no dynamic setup is required. The following would suffice—the second `LineCartesianLayer.Line` would be ignored when only one series is present.

```kt
LineCartesianLayer.LineProvider.series(rememberLine(), rememberLine())
```

In fact, since the `LineCartesianLayer.LineProvider` instantiated by `LineCartesianLayer.LineProvider.series` can go through the `LineCartesianLayer.Line` list repeatedly, and the two `LineCartesianLayer.Line`s here are equal, this would work:

```kt
LineCartesianLayer.LineProvider.series(rememberLine())
```

This is only a passing observation—I’m aware that we’re dealing with a simplified example. There certainly are cases where dynamic, `Transaction`-synchronized `LineCartesianLayer.Line`-related setup is required. Such instances are discussed below.

If we consider a case where there are, let’s say, three series that can be shown and hidden independently, and each series has a unique corresponding `LineCartesianLayer.Line`, we can easily notice that `LineCartesianLayer.LineProvider.series` is off the table. If we provide a static list of three `LineCartesianLayer.Line`s, we may obtain the desired result in some instances (e.g., when the first and second series are shown), but certainly not always (e.g., when first and third series are shown, in which case the third series will be incorrectly linked to the second `LineCartesianLayer.Line`). To ensure correct linkage in all cases, one could assign integer IDs to the three series and use a solution such as that below. Here, the extra associated with `SeriesIDKey` is a list containing the IDs of the visible series. The IDs start at zero.

```kt
val lines =
  listOf(
    rememberLine(/* ... */),
    rememberLine(/* ... */),
    rememberLine(/* ... */),
  )
```

```kt
LineCartesianLayer.LineProvider { seriesIndex, extraStore ->
  lines[extraStore[SeriesIDKey][seriesIndex]]
}
```

It should be noted that with animations enabled, some suboptimal behavior would be observed here. This relates not to `LineCartesianLayer.Line` linkage, which the above solution handles correctly, but rather to `DefaultCartesianLayerDrawingModelInterpolator`, which associates old and new series by index. If, for example, we had all three series shown and hid the second one, then instead of just having the line for the second series disappear, we’d see that for the third one abruptly adopt the shape of that for the second one and then, as the animation progresses, return to the desired shape. We plan on addressing this deficiency, most likely by introducing consumer-specifiable series IDs to help `DefaultCartesianLayerDrawingModelInterpolator` associate series more reliably. (If this approach is taken, the series IDs will likely be passed to `LineCartesianLayer.LineProvider.getLine`, so one will be able to slightly simplify the solution above.) If you’re interested in this enhancement, please feel free to open a feature request.

Finally, I’d like to address cases where having a predefined set of `LineCartesianLayer.Line`s (as above—see `lines`) is insufficient, and one wishes to create `LineCartesianLayer.Line`s on the fly from `LineCartesianLayer.LineProvider.getLine` based on extras. This is possible, but there are currently two shortcomings:

- The output of `LineCartesianLayer.LineProvider.getLine` isn’t cached, so when this approach is taken, equal `LineCartesianLayer.Line`s are created repeatedly. This slightly worsens performance, though in most cases, the difference isn’t noticeable.
- The `compose` module currently lacks a non-composable factory function for `LineCartesianLayer.Line`. Thus, when creating `LineCartesianLayer.Line`s on the fly in Compose, one must use the `LineCartesianLayer.Line` constructor, whose parameters aren’t of Compose types. As a result, one must use `Color.toArgb`, `Dp.value`, and the like, which makes the code slightly less elegant.

These shortcomings are a result of the API not having been designed with on-the-fly `LineCartesianLayer.Line` creation in mind. However, they can be easily addressed, and this is planned. Once again, you’re welcome to open a feature request.
